File: compilers/rjs/transpiler/parser/src/ast/declaration.rs
use crate::ast::expression::{
    Expression,
    Identifier,
    StringLiteral,
    Pattern,
    AccessorType,
    PrivateName,
    PropertyName
};
use crate::ast::StatementListItem;
use super::expression::FunctionExpression;
use serde::{Deserialize, Serialize};
use rjs_attribute_marco::js_node;
use crate::span::Span;

#[js_node]
pub struct VariableDeclaration<'a> {
    pub variant: DeclarationKind,
    pub declarators: Vec<VariableDeclarator<'a>>,
}
#[js_node]
pub enum DeclarationKind {
    Let,
    Const,
    Var,
}
#[js_node]
pub struct VariableDeclarator<'a> {
    pub id: Pattern<'a>,
    pub init: Option<Expression<'a>>,
}
#[js_node]
pub struct FunctionBody<'a> {
    pub body: Vec<StatementListItem<'a>>,
}
#[js_node]
pub struct FunctionDeclaration<'a> {
    pub name: Option<Identifier<'a>>,
    pub params: Vec<Pattern<'a>>,
    pub body: FunctionBody<'a>,
    pub generator: bool,
    pub is_async: bool,
}
#[js_node]
pub struct ClassBody<'a> {
    pub body: Vec<ClassElement<'a>>,
}
#[js_node]
pub struct ClassProperty<'a> {
    pub key: PrivateNameOrPropertyName<'a>,
    pub value: Option<Expression<'a>>,
    pub computed: bool,
    pub short: bool,
    pub is_static: bool,
}
#[js_node]
pub struct ClassMethodDefinition<'a> {
    pub key: PrivateNameOrPropertyName<'a>,
    pub body: FunctionBody<'a>,
    pub params: Vec<Pattern<'a>>,
    pub computed: bool,
    pub generator: bool,
    pub is_async: bool,
    pub is_static: bool,
}
#[js_node]
pub struct ClassAccessor<'a> {
    pub key: PrivateNameOrPropertyName<'a>,
    pub accessor_type: AccessorType,
    pub body: FunctionBody<'a>,
    pub params: Vec<Pattern<'a>>,
    pub computed: bool,
}
#[js_node]
pub struct ClassConstructor<'a> {
    pub key: PrivateNameOrPropertyName<'a>,
    pub body: FunctionBody<'a>,
    pub params: Vec<Pattern<'a>>,
}
#[js_node]
#[serde(tag="type")]
pub enum PrivateNameOrPropertyName<'a> {
    #[serde(rename="PrivateName")]
    Private(PrivateName<'a>),
    #[serde(untagged)]
    Prop(PropertyName<'a>),
}
#[js_node]
#[serde(tag="type")]
pub enum ClassElement<'a> {
    #[serde(rename="ClassProperty")]
    ClassProp(ClassProperty<'a>),
    #[serde(rename="ClassMethodDefintion")]
    ClassMethodDef(ClassMethodDefinition<'a>),
    #[serde(rename="ClassAccessor")]
    ClassAccessor(ClassAccessor<'a>),
    #[serde(rename="ClassConstructor")]
    ClassCtor(ClassConstructor<'a>)
}
#[js_node]
pub struct ClassDeclaration<'a> {
    pub id: Identifier<'a>,
    pub super_class:Option<Expression<'a>>,
    pub body: ClassBody<'a>,
}
#[js_node]
#[serde(tag="type")]
pub enum Declaration<'a> {
    #[serde(rename="FunctionDeclaration")]
    FunDeclar(FunctionDeclaration<'a>),
    #[serde(rename="VariableDeclaration")]
    VarDeclar(VariableDeclaration<'a>),
    #[serde(rename="ClassDeclaration")]
    ClassDeclar(ClassDeclaration<'a>)
}
#[derive(Debug, Clone, PartialEq, Deserialize, Serialize)]
pub struct ImportDeclaration<'a> {
    pub specifiers: Vec<ImportSpecifierKind<'a>>,
    pub source: StringLiteral<'a>
}
#[derive(Debug, Clone, PartialEq, Deserialize, Serialize)]
pub enum ImportSpecifierKind<'a> {
    Default(ImportDefaultSpecifier<'a>),
    Namespace(ImportNamespaceSpecifier<'a>),
    Specifier(ImportSpecifier<'a>),
}
#[derive(Debug, Clone, PartialEq, Deserialize, Serialize)]
pub struct ImportDefaultSpecifier<'a> {
    pub imported: Identifier<'a>
}
#[derive(Debug, Clone, PartialEq, Deserialize, Serialize)]
pub struct ImportNamespaceSpecifier<'a> {
    pub imported: Identifier<'a>
}
#[derive(Debug, Clone, PartialEq, Deserialize, Serialize)]
pub struct ImportSpecifier<'a> {
    pub imported: ModuleName<'a>,
    pub local: Option<Identifier<'a>>
}
#[derive(Debug, Clone, PartialEq, Deserialize, Serialize)]
pub enum ModuleName<'a> {
    Ident(Identifier<'a>),
    String(StringLiteral<'a>),
}
#[derive(Debug, Clone, PartialEq, Deserialize, Serialize)]
pub struct ExportNameDeclaration<'a> {
    pub specifiers: Vec<ExportSpecifier<'a>>,
    pub declaration: Option<Declaration<'a>>,
    pub source: Option<StringLiteral<'a>>,
}
#[derive(Debug, Clone, PartialEq, Deserialize, Serialize)]
pub struct ExportSpecifier<'a> {
   pub exported: ModuleName<'a>,
   pub local: Option<ModuleName<'a>>,
}
#[derive(Debug, Clone, PartialEq, Deserialize, Serialize)]
pub struct ExportDefaultDeclaration<'a> {
    pub declaration: DefaultDeclaration<'a>,
}
#[derive(Debug, Clone, PartialEq, Deserialize, Serialize)]
pub enum DefaultDeclaration<'a> {
    FunctionDeclar(FunctionDeclaration<'a>),
    FunctionExpr(FunctionExpression<'a>),
    ClassDeclar(ClassDeclaration<'a>),
    Expr(Expression<'a>),
}
#[derive(Debug, Clone, PartialEq, Deserialize, Serialize)]
pub struct ExportAllDeclaration<'a> {
    pub exported: Option<Identifier<'a>>,
    pub source: StringLiteral<'a>,
}
#[derive(Debug, Clone, PartialEq, Deserialize, Serialize)]
pub enum ExportDeclaration<'a> {
    Name(ExportNameDeclaration<'a>),
    Default(ExportDefaultDeclaration<'a>),
    All(ExportAllDeclaration<'a>)
}

File: web-infras/common/src/ast/declaration.ts
import { ModuleItem } from "@/src/ast/base";
import { SyntaxKinds } from "@/src/kind";
import { StatementListItem } from "@/src/ast/program";
import { Expression, Identifier, NumberLiteral, PrivateName, StringLiteral } from "@/src/ast/expression";
import { Pattern } from "@/src/ast/pattern";
import {
  TSTypeParameterDeclaration,
  TSTypeAnnotation,
  TSDeclareFunction,
  TSEnumDeclaration,
  TSInterfaceDeclaration,
  TSTypeAliasDeclaration,
} from "@/src/ast/typescript";

export type PropertyName = StringLiteral | NumberLiteral | Identifier | Expression;

export interface VariableDeclaration extends ModuleItem {
  kind: SyntaxKinds.VariableDeclaration;
  declarations: Array<VariableDeclarator>;
  variant: "let" | "const" | "var";
}
export interface VariableDeclarator extends ModuleItem {
  kind: SyntaxKinds.VariableDeclarator;
  id: Pattern;
  init: Expression | null;
}
export interface Function extends Omit<ModuleItem, "kind"> {
  name: Identifier | null;
  params: Array<Pattern>;
  typeParameters: TSTypeParameterDeclaration | undefined;
  returnType: TSTypeAnnotation | undefined;
  body: FunctionBody;
  generator: boolean;
  async: boolean;
}
export interface FunctionBody extends ModuleItem {
  kind: SyntaxKinds.FunctionBody;
  body: Array<StatementListItem>; //TODO: using StatementListItem
}
export interface FunctionDeclaration extends ModuleItem, Function {
  kind: SyntaxKinds.FunctionDeclaration;
  name: Identifier;
}
export interface Class extends Omit<ModuleItem, "kind"> {
  id: Identifier | null;
  superClass: Expression | null;
  body: ClassBody;
  decorators: Decorator[] | null;
}
export interface ClassBody extends ModuleItem {
  kind: SyntaxKinds.ClassBody;
  body: Array<ClassElement>;
}
export type ClassElementName = PropertyName | PrivateName;
export interface ClassProperty extends ModuleItem {
  kind: SyntaxKinds.ClassProperty;
  decorators: Decorator[] | null;
  // JS modifier
  static: boolean;
  // ClassElementName
  key: ClassElementName;
  computed: boolean;
  // TS Modifier
  accessibility: "private" | "public" | "protected" | null;
  abstract: boolean;
  // TS type
  optional: boolean;
  typeAnnotation: TSTypeAnnotation | undefined;
  // Default value
  value: Expression | undefined; // actually is assignment expression,
}
export interface ClassAccessorProperty extends ModuleItem {
  kind: SyntaxKinds.ClassAccessorProperty;
  decorators: Decorator[] | null;
  // JS modifier
  static: boolean;
  // ClassElementName
  key: ClassElementName;
  computed: boolean;
  // TS Modifier
  accessibility: "private" | "public" | "protected" | null;
  abstract: boolean;
  // TS type
  typeAnnotation: TSTypeAnnotation | undefined;
  // Default value
  value: Expression | undefined; // actually is assignment expression,
}

export interface ClassConstructor extends ModuleItem {
  kind: SyntaxKinds.ClassConstructor;
  // ClassElementName
  key: Identifier;
  // function
  params: Array<Pattern>;
  body: FunctionBody;
  // TS type
  returnType: TSTypeAnnotation | undefined;
  // TS modifier
  accessibility: "private" | "public" | "protected" | null;
}
export interface TSClassConstructorDeclaration extends ModuleItem {
  kind: SyntaxKinds.TSClassConstructorDeclaration;
  // ClassElementName
  key: Identifier;
  // function
  params: Array<Pattern>;
  // TS Type
  returnTye: TSTypeAnnotation | undefined;
  // TS modifier
  accessibility: "private" | "public" | "protected" | null;
}
export interface ClassMethodDefinition extends ModuleItem {
  kind: SyntaxKinds.ClassMethodDefinition;
  decorators: Decorator[] | null;
  // JS modifier
  generator: boolean;
  async: boolean;
  static: boolean;
  // ClassElementName
  key: ClassElementName;
  computed: boolean;
  // function
  params: Array<Pattern>;
  body: FunctionBody;
  // TS type
  typeParameters: TSTypeParameterDeclaration | undefined;
  returnType: TSTypeAnnotation | undefined;
  // TS modifier
  optional: boolean;
  accessibility: "private" | "public" | "protected" | null;
}
export interface ClassMethodDeclaration extends ModuleItem {
  kind: SyntaxKinds.ClassMethodDeclaration;
  decorators: Decorator[] | null;
  // JS modifier
  generator: boolean;
  async: boolean;
  static: boolean;
  // ClassElementName
  key: ClassElementName;
  computed: boolean;
  // function
  params: Array<Pattern>;
  // TS type
  typeParameters: TSTypeParameterDeclaration | undefined;
  returnType: TSTypeAnnotation | undefined;
  // TS modifier
  optional: boolean;
  accessibility: "private" | "public" | "protected" | null;
  abstract: boolean;
}
export interface ClassAccessor extends ModuleItem {
  kind: SyntaxKinds.ClassAccessor;
  decorators: Decorator[] | null;
  // JS modifier
  type: "get" | "set";
  static: boolean;
  // ClassElementName
  key: ClassElementName;
  computed: boolean;
  // function
  params: Array<Pattern>;
  body: FunctionBody;
  // TS type
  typeParameters: TSTypeParameterDeclaration | undefined;
  returnType: TSTypeAnnotation | undefined;
  // TS modifier
  accessibility: "private" | "public" | "protected" | null;
}
// NOTE: accessor must be abstract to be declaration.
export interface TSAbstractClassAccessorDeclaration extends ModuleItem {
  kind: SyntaxKinds.TSAbstractClassAccessorDeclaration;
  decorators: Decorator[] | null;
  // JS modifier
  type: "get" | "set";
  static: boolean;
  // ClassElementName
  key: ClassElementName;
  computed: boolean;
  // function
  params: Array<Pattern>;
  // TS type
  typeParameters: TSTypeParameterDeclaration | undefined;
  returnType: TSTypeAnnotation | undefined;
  // TS modifier
  accessibility: "private" | "public" | "protected" | null;
}
export interface ClassStaticBlock extends ModuleItem {
  kind: SyntaxKinds.ClassStaticBlock;
  body: Array<StatementListItem>;
}
export type ClassElement =
  | ClassProperty
  | ClassAccessorProperty
  | ClassMethodDefinition
  | ClassMethodDeclaration
  | ClassAccessor
  | TSAbstractClassAccessorDeclaration
  | ClassConstructor
  | ClassStaticBlock;
export interface ClassDeclaration extends ModuleItem, Class {
  kind: SyntaxKinds.ClassDeclaration;
}
export interface Decorator extends ModuleItem {
  kind: SyntaxKinds.Decorator;
  expression: Expression;
}
export type Declaration = FunctionDeclaration | VariableDeclaration | ClassDeclaration | TSDeclaration;

export type TSDeclaration =
  | TSTypeAliasDeclaration
  | TSInterfaceDeclaration
  | TSEnumDeclaration
  | TSDeclareFunction;
